// axentix/dist/js/axentix.esm.js
export { Axentix };"use strict";

/**
 * Class AxentixComponent
 * @class
 */
class AxentixComponent {
  /**
   * Sync all listeners
   */
  sync() {
    Axentix.createEvent(this.el, 'component.sync');

    this._removeListeners();

    this._setupListeners();
  }
  /**
   * Reset component
   */


  reset() {
    Axentix.createEvent(this.el, 'component.reset');

    this._removeListeners();

    this._setup();
  }

}
/**
 * Class Axentix
 * @class
 */


class Axentix {
  /**
   * Construct Axentix instance
   * @constructor
   * @param {String} component
   * @param {Object} options
   */
  constructor(component, options) {
    this.component = component[0].toUpperCase() + component.slice(1).toLowerCase();
    this.isAll = component === 'all' ? true : false;
    this.options = this.isAll ? {} : options;

    this._init();
  }
  /**
   * Init components
   */


  _init() {
    var componentList = Axentix.Config.getAutoInitElements();
    var isInList = componentList.hasOwnProperty(this.component);

    if (isInList) {
      var ids = this._detectIds(componentList[this.component]);

      this._instanciate(ids, this.component);
    } else if (this.isAll) {
      Object.keys(componentList).map(component => {
        var ids = this._detectIds(componentList[component]);

        ids.length > 0 ? this._instanciate(ids, component) : '';
      });
    }
  }
  /**
   * Detects all ids
   * @param {NodeListOf<Element>} component
   * @return {Array<String>}
   */


  _detectIds(component) {
    var idList = [];
    component.forEach(el => {
      idList.push('#' + el.id);
    });
    return idList;
  }
  /**
   * Instanciate components
   * @param {Array<String>} ids
   * @param {String} component
   */


  _instanciate(ids, component) {
    ids.map(id => {
      var constructor = Axentix[component];
      var args = [id, this.options];

      try {
        new constructor(...args);
      } catch (error) {
        console.error('Axentix : Unable to load ' + component);
      }
    });
  }

}

Axentix.instances = []; // require: ax-core.js

Axentix.Config = (() => {
  var config = {
    components: [],
    plugins: []
  };

  var get = () => {
    return config;
  };

  var getDataElements = () => {
    var dataComponents = config.components.filter(component => component.dataDetection);
    var dataPlugins = config.plugins.filter(plugin => plugin.dataDetection);
    return [...dataComponents, ...dataPlugins].reduce((acc, el) => {
      acc.push(el.name);
      return acc;
    }, []);
  };

  var getAutoInitElements = () => {
    var autoInitComponents = config.components.filter(component => component.autoInit && component.autoInit.enabled);
    var autoInitPlugins = config.plugins.filter(plugin => plugin.autoInit && plugin.autoInit.enabled);
    return [...autoInitComponents, ...autoInitPlugins].reduce((acc, el) => {
      acc[el.name] = document.querySelectorAll(el.autoInit.selector);
      return acc;
    }, {});
  };
  /**
   * Register an element
   * @param {{ name: string, dataDetection?: boolean, autoInit?: {enabled: boolean, selector: string}, class: any }} el
   * @param {String} term
   */


  var register = (el, term) => {
    if (!el.name || !el.class) {
      console.error("Error registering ".concat(term, " : Missing required parameters."));
      return;
    }

    if (config[term].some(elem => elem.name === el.name)) {
      console.error("Error registering ".concat(term, " : Already exist."));
      return;
    }

    config[term].push(el);
    Axentix[el.name] = el.class;
  };

  var registerComponent = component => {
    register(component, 'components');
  };

  var registerPlugin = plugin => {
    register(plugin, 'plugins');
  };

  return {
    get,
    getDataElements,
    getAutoInitElements,
    registerComponent,
    registerPlugin
  };
})();

Axentix.DataDetection = (() => {
  var getFormattedName = name => {
    return name.replace(/[\w]([A-Z])/g, s => {
      return s[0] + '-' + s[1];
    }).toLowerCase();
  };

  var getName = function getName(name) {
    var baseName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
    var fmtName = getFormattedName(name);
    return baseName ? baseName + '-' + fmtName : fmtName;
  };

  var getOptions = function getOptions(obj, component, element) {
    var baseName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '';
    return Object.keys(obj).reduce((acc, name, i) => {
      if (typeof obj[name] === 'object' && obj[name] !== null) {
        var tmpOptName = name[0].toUpperCase() + name.slice(1).toLowerCase();
        Axentix.Config.getDataElements().includes(tmpOptName) && component !== 'Collapsible' && tmpOptName !== 'Sidenav' ? obj[name] = Axentix[tmpOptName].getDefaultOptions() : '';
        var fmtName = baseName ? baseName + '-' + name : name;
        var keys = getOptions(obj[name], component, element, fmtName);
        Object.keys(keys).length === 0 && obj.constructor === Object ? '' : acc[name] = keys;
      } else if (obj[name] !== null) {
        var dataAttribute = 'data-' + component.toLowerCase() + '-' + getName(name, baseName);

        if (element.hasAttribute(dataAttribute)) {
          var attr = element.getAttribute(dataAttribute);
          acc[name] = typeof obj[name] === 'boolean' ? attr === 'true' : typeof obj[name] === 'number' ? Number(attr) : attr;
        }
      }

      return acc;
    }, {});
  };
  /**
   * Format options provided
   * @param {string} component
   * @param {Element} element
   * @return {object}
   */


  var formatOptions = (component, element) => {
    var defaultOptions = Axentix[component].getDefaultOptions();
    return getOptions(defaultOptions, component, element);
  };

  var setup = () => {
    var elements = document.querySelectorAll('[data-ax]');
    elements.forEach(el => {
      var component = el.dataset.ax;
      component = component[0].toUpperCase() + component.slice(1).toLowerCase();

      if (!Axentix.Config.getDataElements().includes(component)) {
        console.error("Error: This component doesn't exist.", el);
        return;
      }

      var options = formatOptions(component, el);

      try {
        new Axentix[component]("#".concat(el.id), options, true);
      } catch (error) {
        console.error('Axentix Data: Unable to load ' + component);
      }
    });
  };

  var setupAll = () => {
    new Axentix('all');
  };

  return {
    setup,
    setupAll,
    formatOptions
  };
})();

document.addEventListener('DOMContentLoaded', () => {
  document.documentElement.dataset.axentix ? Axentix.DataDetection.setupAll() : '';
  Axentix.DataDetection.setup();
});

(() => {
  /**
   * Class Caroulix
   * @class
   */
  class Caroulix extends AxentixComponent {
    static getDefaultOptions() {
      return {
        fixedHeight: true,
        height: '',
        animationDuration: 500,
        animationType: 'slide',
        indicators: {
          enabled: false,
          isFlat: false,
          customClasses: ''
        },
        autoplay: {
          enabled: true,
          interval: 5000,
          side: 'right'
        }
      };
    }
    /**
     * Construct Caroulix instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Caroulix',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Caroulix', options, this.el, isLoadedWithData);

      this._setup();
    }

    _setup() {
      Axentix.createEvent(this.el, 'caroulix.setup');
      var animationList = ['slide'];
      animationList.includes(this.options.animationType) ? '' : this.options.animationType = 'slide';
      var autoplaySides = ['right', 'left'];
      autoplaySides.includes(this.options.autoplay.side) ? '' : this.options.autoplay.side = 'right';
      this.currentItemIndex = 0;
      this.isAnimated = false;
      this.animFunction = '_animation' + this.options.animationType.charAt(0).toUpperCase() + this.options.animationType.substring(1);

      this._getChildrens();

      this.options.indicators.enabled ? this._enableIndicators() : '';

      this._getActiveElementIndex();

      this._setupListeners();

      this.el.classList.add('anim-' + this.options.animationType);
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.windowResizeRef = this._handleResizeEvent.bind(this);
      window.addEventListener('resize', this.windowResizeRef);

      if (this.arrowPrev && this.arrowNext) {
        this.arrowPrevRef = this.prev.bind(this, 1);
        this.arrowNextRef = this.next.bind(this, 1);
        this.arrowPrev.addEventListener('click', this.arrowPrevRef);
        this.arrowNext.addEventListener('click', this.arrowNextRef);
      }
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      window.removeEventListener('resize', this.windowResizeRef);
      this.windowResizeRef = undefined;

      if (this.arrowPrev && this.arrowNext) {
        this.arrowPrev.removeEventListener('click', this.arrowPrevRef);
        this.arrowNext.removeEventListener('click', this.arrowNextRef);
        this.arrowPrevRef = undefined;
        this.arrowNextRef = undefined;
      }
    }
    /**
     * Handle resize event
     */


    _handleResizeEvent(e) {
      this.updateHeight();
    }
    /**
     * Get caroulix childrens
     */


    _getChildrens() {
      this.childrens = Array.from(this.el.children).reduce((acc, child) => {
        child.classList.contains('caroulix-item') ? acc.push(child) : '';
        child.classList.contains('caroulix-prev') ? this.arrowPrev = child : '';
        child.classList.contains('caroulix-next') ? this.arrowNext = child : '';
        return acc;
      }, []);
    }

    _getActiveElementIndex() {
      this.childrens.map((child, i) => {
        if (child.classList.contains('active')) {
          this.currentItemIndex = i;
        }
      });
      var item = this.childrens[this.currentItemIndex];
      item.classList.contains('active') ? '' : item.classList.add('active');
      this.options.indicators.enabled ? this.indicators.children[this.currentItemIndex].classList.add('active') : '';

      this._waitUntilLoad(item);
    }

    _waitUntilLoad(item) {
      var isImage = false;

      if (this.options.fixedHeight) {
        this.totalLoadChild = 0;
        this.totalLoadedChild = 0;
        this.childrens.map(child => {
          var waitItem = child.querySelector('img, video');

          if (waitItem) {
            isImage = true;
            this.totalLoadChild++;

            if (waitItem.complete) {
              this._initWhenLoaded(waitItem, true);
            } else {
              waitItem.loadRef = this._initWhenLoaded.bind(this, waitItem);
              waitItem.addEventListener('load', waitItem.loadRef);
            }
          }
        });
      } else {
        var childItem = item.querySelector('img, video');

        if (childItem) {
          isImage = true;

          if (childItem.complete) {
            this._initWhenLoaded(childItem, true);
          } else {
            childItem.loadRef = this._initWhenLoaded.bind(this, childItem);
            childItem.addEventListener('load', childItem.loadRef);
          }
        }
      }

      if (!isImage) {
        this.updateHeight();
        this.options.autoplay.enabled ? this.play() : '';
      }
    }
    /**
     * Update height & remove listener when active element is loaded
     * @param {Element} item
     * @param {Boolean} alreadyLoad
     */


    _initWhenLoaded(item, alreadyLoad) {
      if (this.options.fixedHeight) {
        if (!alreadyLoad) {
          item.removeEventListener('load', item.loadRef);
          item.loadRef = undefined;
        }

        this.totalLoadedChild++;

        if (this.totalLoadedChild === this.totalLoadChild) {
          this.updateHeight();
          this.totalLoadedChild = undefined;
          this.totalLoadChild = undefined;
          this.options.autoplay.enabled ? this.play() : '';
        }
      } else {
        this.updateHeight();
        item.removeEventListener('load', item.loadRef);
        item.loadRef = undefined;
        this.options.autoplay.enabled ? this.play() : '';
      }
    }

    _setMaxHeight() {
      if (this.options.height) {
        this.el.style.height = this.options.height;
        return;
      }

      var childrensHeight = this.childrens.map(child => {
        return child.offsetHeight;
      });
      this.maxHeight = Math.max(...childrensHeight);
      this.el.style.height = this.maxHeight + 'px';
    }
    /**
     * Dynamic height option
     * @param {number} index
     */


    _setDynamicHeight() {
      var index = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.currentItemIndex;
      var height = this.childrens[index].offsetHeight;
      this.el.style.height = height + 'px';
    }
    /**
     * Enable indicators
     */


    _enableIndicators() {
      this.indicators = document.createElement('ul');
      this.indicators.classList.add('caroulix-indicators');
      this.options.indicators.isFlat ? this.indicators.classList.add('caroulix-flat') : '';
      this.options.indicators.customClasses ? this.indicators.className = this.indicators.className + ' ' + this.options.indicators.customClasses : '';

      for (var i = 0; i < this.childrens.length; i++) {
        var li = document.createElement('li');
        li.triggerRef = this._handleIndicatorClick.bind(this, i);
        li.addEventListener('click', li.triggerRef);
        this.indicators.appendChild(li);
      }

      this.el.appendChild(this.indicators);
    }
    /***** Animation Section *****/

    /**
     * Slide animation
     * @param {number} number
     * @param {string} side
     */


    _animationSlide(number, side) {
      var nextItem = this.childrens[number];
      var currentItem = this.childrens[this.currentItemIndex];
      var nextItemPercentage = '',
          currentItemPercentage = '';

      if (side === 'right') {
        nextItemPercentage = '100%';
        currentItemPercentage = '-100%';
      } else {
        nextItemPercentage = '-100%';
        currentItemPercentage = '100%';
      }

      nextItem.style.transform = "translateX(".concat(nextItemPercentage, ")");
      nextItem.classList.add('active');
      setTimeout(() => {
        nextItem.style.transitionDuration = this.options.animationDuration + 'ms';
        nextItem.style.transform = '';
        currentItem.style.transitionDuration = this.options.animationDuration + 'ms';
        currentItem.style.transform = "translateX(".concat(currentItemPercentage, ")");
      }, 50);
      setTimeout(() => {
        nextItem.removeAttribute('style');
        currentItem.classList.remove('active');
        currentItem.removeAttribute('style');
        this.currentItemIndex = number;
        this.isAnimated = false;
        this.options.autoplay.enabled ? this.play() : '';
      }, this.options.animationDuration + 50);
    }
    /***** [END] Animation Section [END] *****/

    /**
     * Handle indicator click
     * @param {number} i
     * @param {Event} e
     */


    _handleIndicatorClick(i, e) {
      e.preventDefault();

      if (i === this.currentItemIndex) {
        return;
      }

      this.goTo(i);
    }

    _getPreviousItemIndex(step) {
      var previousItemIndex = 0;
      var index = this.currentItemIndex;

      for (var i = 0; i < step; i++) {
        if (index > 0) {
          previousItemIndex = index - 1;
          index--;
        } else {
          index = this.childrens.length - 1;
          previousItemIndex = index;
        }
      }

      return previousItemIndex;
    }

    _getNextItemIndex(step) {
      var nextItemIndex = 0;
      var index = this.currentItemIndex;

      for (var i = 0; i < step; i++) {
        if (index < this.childrens.length - 1) {
          nextItemIndex = index + 1;
          index++;
        } else {
          index = 0;
          nextItemIndex = index;
        }
      }

      return nextItemIndex;
    }
    /**
     * Update height of caroulix container
     * @param {number} indexRef
     */


    updateHeight(indexRef) {
      this.options.fixedHeight ? this._setMaxHeight() : this._setDynamicHeight(indexRef);
    }
    /**
     * Go to {n} item
     * @param {number} number
     * @param {string} side
     */


    goTo(number, side) {
      if (this.isAnimated || number === this.currentItemIndex) {
        return;
      }

      side ? '' : number > this.currentItemIndex ? side = 'right' : side = 'left';
      this.options.autoplay.enabled && this.autoTimeout ? this.stop() : '';
      Axentix.createEvent(this.el, 'caroulix.slide', {
        side,
        nextElement: this.childrens[number],
        currentElement: this.childrens[this.currentItemIndex]
      });
      this.isAnimated = true;

      if (this.options.indicators.enabled) {
        Array.from(this.indicators.children).map(li => {
          li.removeAttribute('class');
        });
        this.indicators.children[number].classList.add('active');
      }

      this.options.fixedHeight ? '' : this.updateHeight(number);
      this[this.animFunction](number, side);
    }

    prev() {
      var step = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

      if (this.isAnimated) {
        return;
      }

      Axentix.createEvent(this.el, 'caroulix.prev', {
        step
      });

      var previousItemIndex = this._getPreviousItemIndex(step);

      this.goTo(previousItemIndex, 'left');
    }

    next() {
      var step = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

      if (this.isAnimated) {
        return;
      }

      Axentix.createEvent(this.el, 'caroulix.next', {
        step
      });

      var nextItemIndex = this._getNextItemIndex(step);

      this.goTo(nextItemIndex, 'right');
    }

    play() {
      this.autoTimeout = setTimeout(() => {
        this.options.autoplay.side === 'right' ? this.next() : this.prev();
      }, this.options.autoplay.interval);
    }

    stop() {
      clearTimeout(this.autoTimeout);
      this.autoTimeout = false;
    }

  }

  Axentix.Config.registerComponent({
    class: Caroulix,
    name: 'Caroulix',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.caroulix:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Collapsible
   * @class
   */
  class Collapsible extends AxentixComponent {
    static getDefaultOptions() {
      return {
        animationDuration: 300,
        sidenav: {
          activeClass: true,
          activeWhenOpen: true,
          autoClose: true
        }
      };
    }
    /**
     * Construct Collapsible instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Collapsible',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Collapsible', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'collapsible.setup');
      this.collapsibleTriggers = document.querySelectorAll('.collapsible-trigger');
      this.isInitialStart = true;
      this.isActive = this.el.classList.contains('active') ? true : false;
      this.isAnimated = false;
      this.isInSidenav = false;
      this.childIsActive = false;

      this._setupListeners();

      this.el.style.transitionDuration = this.options.animationDuration + 'ms';

      this._detectSidenav();

      this._detectChild();

      this.options.sidenav.activeClass ? this._addActiveInSidenav() : '';
      this.isActive ? this.open() : '';
      this.isInitialStart = false;
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.listenerRef = this._onClickTrigger.bind(this);
      this.collapsibleTriggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          trigger.addEventListener('click', this.listenerRef);
        }
      });
      this.resizeRef = this._handleResize.bind(this);
      window.addEventListener('resize', this.resizeRef);
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      this.collapsibleTriggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          trigger.removeEventListener('click', this.listenerRef);
        }
      });
      this.listenerRef = undefined;
      window.removeEventListener('resize', this.resizeRef);
      this.resizeRef = undefined;
    }
    /**
     * Reset collapsible maxHeight
     */


    _handleResize() {
      this.isActive && !this.isInSidenav ? this.el.style.maxHeight = this.el.scrollHeight + 'px' : '';
    }
    /**
     * Check if collapsible is in sidenav
     */


    _detectSidenav() {
      var sidenavElem = this.el.closest('.sidenav');

      if (sidenavElem) {
        this.isInSidenav = true;
        this.sidenavId = sidenavElem.id;
      }
    }
    /**
     * Check if collapsible have active childs
     */


    _detectChild() {
      for (var child of this.el.children) {
        if (child.classList.contains('active')) {
          this.childIsActive = true;
          break;
        }
      }
    }
    /**
     * Add active class to trigger and collapsible
     */


    _addActiveInSidenav() {
      if (this.childIsActive && this.isInSidenav) {
        var triggers = document.querySelectorAll('.sidenav .collapsible-trigger');
        triggers.forEach(trigger => {
          if (trigger.dataset.target === this.el.id) {
            trigger.classList.add('active');
          }
        });
        this.el.classList.add('active');
        this.open();
        this.isActive = true;
      }
    }
    /**
     * Enable / disable active state to trigger when collapsible is in sidenav
     * @param {boolean} state enable or disable
     */


    _addActiveToTrigger(state) {
      var triggers = document.querySelectorAll('.sidenav .collapsible-trigger');
      triggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          state ? trigger.classList.add('active') : trigger.classList.remove('active');
        }
      });
    }
    /**
     * Auto close others collapsible
     */


    _autoClose() {
      if (!this.isInitialStart && this.isInSidenav) {
        Axentix.getInstanceByType('Collapsible').map(collapsible => {
          if (collapsible.isInSidenav && collapsible.sidenavId === this.sidenavId && collapsible.el.id !== this.el.id) {
            collapsible.close();
          }
        });
      }
    }
    /**
     * Apply overflow hidden and automatically remove
     */


    _applyOverflow() {
      this.el.style.overflow = 'hidden';
      setTimeout(() => {
        this.el.style.overflow = '';
      }, this.options.animationDuration);
    }
    /**
     * Handle click on trigger
     * @param {Event} e
     */


    _onClickTrigger(e) {
      e.preventDefault();

      if (this.isAnimated) {
        return;
      }

      this.isActive ? this.close() : this.open();
    }
    /**
     * Open collapsible
     */


    open() {
      if (this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'collapsible.open');
      this.isActive = true;
      this.isAnimated = true;
      this.el.style.display = 'block';

      this._applyOverflow();

      this.el.style.maxHeight = this.el.scrollHeight + 'px';
      this.options.sidenav.activeWhenOpen ? this._addActiveToTrigger(true) : '';
      this.options.sidenav.autoClose ? this._autoClose() : '';
      setTimeout(() => {
        this.isAnimated = false;
      }, this.options.animationDuration);
    }
    /**
     * Close collapsible
     */


    close() {
      if (!this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'collapsible.close');
      this.isAnimated = true;
      this.el.style.maxHeight = '';

      this._applyOverflow();

      this.options.sidenav.activeWhenOpen ? this._addActiveToTrigger(false) : '';
      setTimeout(() => {
        this.el.style.display = '';
        this.isAnimated = false;
        this.isActive = false;
      }, this.options.animationDuration);
    }

  }

  Axentix.Config.registerComponent({
    class: Collapsible,
    name: 'Collapsible',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.collapsible:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Dropdown
   * @class
   */
  class Dropdown extends AxentixComponent {
    static getDefaultOptions() {
      return {
        animationDuration: 300,
        animationType: 'none',
        hover: false,
        autoClose: true,
        preventViewport: false
      };
    }
    /**
     * Construct Dropdown instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Dropdown',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Dropdown', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'dropdown.setup');
      this.dropdownContent = this.el.querySelector('.dropdown-content');
      this.dropdownTrigger = this.el.querySelector('.dropdown-trigger');
      this.isAnimated = false;
      this.isActive = this.el.classList.contains('active') ? true : false;
      this.options.hover ? this.el.classList.add('active-hover') : this._setupListeners();
      this.options.preventViewport ? this.el.classList.add('dropdown-vp') : '';

      this._setupAnimation();
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      if (this.options.hover) {
        return;
      }

      this.listenerRef = this._onClickTrigger.bind(this);
      this.dropdownTrigger.addEventListener('click', this.listenerRef);
      this.documentClickRef = this._onDocumentClick.bind(this);
      document.addEventListener('click', this.documentClickRef, true);
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      if (this.options.hover) {
        return;
      }

      this.dropdownTrigger.removeEventListener('click', this.listenerRef);
      this.listenerRef = undefined;
      document.removeEventListener('click', this.documentClickRef, true);
      this.documentClickRef = undefined;
    }
    /**
     * Check and init animation
     */


    _setupAnimation() {
      var animationList = ['none', 'fade'];
      this.options.animationType = this.options.animationType.toLowerCase();
      animationList.includes(this.options.animationType) ? '' : this.options.animationType = 'none';

      if (this.options.animationType !== 'none' && !this.options.hover) {
        if (this.options.hover) {
          this.el.style.animationDuration = this.options.animationDuration + 'ms';
        } else {
          this.el.style.transitionDuration = this.options.animationDuration + 'ms';
        }

        this.el.classList.add('anim-' + this.options.animationType);
      }
    }
    /**
     * Handle click on document click
     */


    _onDocumentClick(e) {
      if (e.target.matches('.dropdown-trigger')) {
        return;
      }

      if (this.isAnimated || !this.isActive) {
        return;
      }

      this.close();
    }
    /**
     * Handle click on trigger
     */


    _onClickTrigger(e) {
      e.preventDefault();

      if (this.isAnimated) {
        return;
      }

      this.isActive ? this.close() : this.open();
    }

    _autoClose() {
      Axentix.getInstanceByType('Dropdown').map(dropdown => {
        dropdown.el.id !== this.el.id ? dropdown.close() : '';
      });
    }

    _setContentHeight() {
      var elRect = this.dropdownContent.getBoundingClientRect();
      var bottom = elRect.height - (elRect.bottom - (window.innerHeight || document.documentElement.clientHeight)) - 10;
      this.dropdownContent.style.maxHeight = bottom + 'px';
    }
    /**
     * Open dropdown
     */


    open() {
      if (this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'dropdown.open');
      this.dropdownContent.style.display = 'flex';
      this.options.preventViewport ? this._setContentHeight() : '';
      setTimeout(() => {
        this.el.classList.add('active');
        this.isActive = true;
      }, 10);
      this.options.autoClose ? this._autoClose() : '';

      if (this.options.animationType !== 'none') {
        this.isAnimated = true;
        setTimeout(() => {
          this.isAnimated = false;
          Axentix.createEvent(this.el, 'dropdown.opened');
        }, this.options.animationDuration);
      } else {
        Axentix.createEvent(this.el, 'dropdown.opened');
      }
    }
    /**
     * Close dropdown
     */


    close() {
      if (!this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'dropdown.close');
      this.el.classList.remove('active');

      if (this.options.animationType !== 'none') {
        this.isAnimated = true;
        setTimeout(() => {
          this.dropdownContent.style.display = '';
          this.isAnimated = false;
          this.isActive = false;
          Axentix.createEvent(this.el, 'dropdown.closed');
        }, this.options.animationDuration);
      } else {
        this.dropdownContent.style.display = '';
        this.isAnimated = false;
        this.isActive = false;
        Axentix.createEvent(this.el, 'dropdown.closed');
      }
    }

  }

  Axentix.Config.registerComponent({
    class: Dropdown,
    name: 'Dropdown',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.dropdown:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Fab
   * @class
   */
  class Fab extends AxentixComponent {
    static getDefaultOptions() {
      return {
        animationDuration: 300,
        hover: true,
        direction: 'top',
        position: 'bottom-right',
        offsetX: '1rem',
        offsetY: '1.5rem'
      };
    }
    /**
     * Construct Fab instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Fab',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Fab', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'fab.setup');
      this.isAnimated = false;
      this.isActive = false;
      this.trigger = document.querySelector('#' + this.el.id + ' .fab-trigger');
      this.fabMenu = document.querySelector('#' + this.el.id + ' .fab-menu');

      this._verifOptions();

      this._setupListeners();

      this.el.style.transitionDuration = this.options.animationDuration + 'ms';

      this._setProperties();
    }
    /**
     * Options check
     */


    _verifOptions() {
      var directionList = ['right', 'left', 'top', 'bottom'];
      directionList.includes(this.options.direction) ? '' : this.options.direction = 'top';
      var positionList = ['top-right', 'top-left', 'bottom-right', 'bottom-left'];
      positionList.includes(this.options.position) ? '' : this.options.position = 'bottom-right';
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      if (this.options.hover) {
        this.openRef = this.open.bind(this);
        this.closeRef = this.close.bind(this);
        this.el.addEventListener('mouseenter', this.openRef);
        this.el.addEventListener('mouseleave', this.closeRef);
      } else {
        this.listenerRef = this._onClickTrigger.bind(this);
        this.el.addEventListener('click', this.listenerRef);
      }

      this.documentClickRef = this._handleDocumentClick.bind(this);
      document.addEventListener('click', this.documentClickRef, true);
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      if (this.options.hover) {
        this.el.removeEventListener('mouseenter', this.openRef);
        this.el.removeEventListener('mouseleave', this.closeRef);
        this.openRef = undefined;
        this.closeRef = undefined;
      } else {
        this.el.removeEventListener('click', this.listenerRef);
        this.listenerRef = undefined;
      }

      document.removeEventListener('click', this.documentClickRef, true);
      this.documentClickRef = undefined;
    }
    /**
     * Set different options on element
     */


    _setProperties() {
      this.options.position.split('-')[0] === 'top' ? this.el.style.top = this.options.offsetY : this.el.style.bottom = this.options.offsetY;
      this.options.position.split('-')[1] === 'right' ? this.el.style.right = this.options.offsetX : this.el.style.left = this.options.offsetX;
      this.options.direction === 'top' || this.options.direction === 'bottom' ? '' : this.el.classList.add('fab-dir-x');

      this._setMenuPosition();
    }
    /**
     * Set menu position
     */


    _setMenuPosition() {
      if (this.options.direction === 'top' || this.options.direction === 'bottom') {
        var height = this.trigger.clientHeight;
        this.options.direction === 'top' ? this.fabMenu.style.bottom = height + 'px' : this.fabMenu.style.top = height + 'px';
      } else {
        var width = this.trigger.clientWidth;
        this.options.direction === 'right' ? this.fabMenu.style.left = width + 'px' : this.fabMenu.style.right = width + 'px';
      }
    }
    /**
     * Handle document click event
     * @param {Event} e
     */


    _handleDocumentClick(e) {
      var isInside = this.el.contains(e.target);
      !isInside && this.isActive ? this.close() : '';
    }
    /**
     * Handle click on trigger
     * @param {Event} e
     */


    _onClickTrigger(e) {
      e.preventDefault();

      if (this.isAnimated) {
        return;
      }

      this.isActive ? this.close() : this.open();
    }
    /**
     * Open fab
     */


    open() {
      if (this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'fab.open');
      this.isAnimated = true;
      this.isActive = true;
      this.el.classList.add('active');
      setTimeout(() => {
        this.isAnimated = false;
      }, this.options.animationDuration);
    }
    /**
     * Close fab
     */


    close() {
      if (!this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'fab.close');
      this.isAnimated = true;
      this.isActive = false;
      this.el.classList.remove('active');
      setTimeout(() => {
        this.isAnimated = false;
      }, this.options.animationDuration);
    }

  }

  Axentix.Config.registerComponent({
    class: Fab,
    name: 'Fab',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.fab:not(i):not(.no-axentix-init)'
    }
  });
})();

Axentix.Forms = (() => {
  var isInit = true;
  /**
   * Detect attribute & state of all inputs
   * @param {NodeListOf<Element>} inputElements
   */

  var detectAllInputs = inputElements => {
    inputElements.forEach(input => detectInput(input));
  };
  /**
   * Delay detection of all inputs
   * @param {NodeListOf<Element>} inputElements
   */


  var delayDetectionAllInputs = inputElements => {
    if (isInit) {
      isInit = false;
      return;
    }

    setTimeout(() => {
      detectAllInputs(inputElements);
    }, 10);
  };
  /**
   * Detect attribute & state of an input
   * @param {Element} input
   */


  var detectInput = input => {
    var isActive = input.parentElement.classList.contains('active');
    var hasContent = input.value.length > 0 || input.tagName !== 'SELECT' && input.placeholder.length > 0 || input.tagName === 'SELECT' || input.matches('[type="date"]') || input.matches('[type="month"]') || input.matches('[type="week"]') || input.matches('[type="time"]');
    var isFocused = document.activeElement === input;
    var isDisabled = input.hasAttribute('disabled') || input.hasAttribute('readonly');

    if (input.firstInit) {
      updateInput(input, isActive, hasContent, isFocused);
      input.firstInit = false;
      input.isInit = true;
    } else {
      isDisabled ? '' : updateInput(input, isActive, hasContent, isFocused);
    }
  };
  /**
   * Update input field
   * @param {Element} input
   * @param {boolean} isActive
   * @param {boolean} hasContent
   * @param {boolean} isFocused
   */


  var updateInput = (input, isActive, hasContent, isFocused) => {
    var isTextArea = input.type === 'textarea';

    if (!isActive && (hasContent || isFocused)) {
      input.parentElement.classList.add('active');
    } else if (isActive && !(hasContent || isFocused)) {
      input.parentElement.classList.remove('active');
    }

    isTextArea ? '' : setFormPosition(input);
    isFocused && !isTextArea ? input.parentElement.classList.add('is-focused') : input.parentElement.classList.remove('is-focused');
    isFocused && isTextArea ? input.parentElement.classList.add('is-txtarea-focused') : input.parentElement.classList.remove('is-txtarea-focused');
  };
  /**
   * Add bottom position variable to form
   * @param {Element} input
   */


  var setFormPosition = input => {
    var style = window.getComputedStyle(input.parentElement);
    var height = parseFloat(input.clientHeight),
        padding = parseFloat(style.paddingTop),
        border = parseFloat(style.borderTopWidth);
    var pos = padding + border + height + 'px';
    input.parentElement.style.setProperty('--form-material-position', pos);
  };
  /**
   * Handle listeners
   * @param {NodeListOf<Element>} inputs
   * @param {Event} e
   */


  var handleListeners = (inputs, e) => {
    inputs.forEach(input => {
      input === e.target ? detectInput(input) : '';
    });
  };
  /**
   * Handle form reset event
   * @param {NodeListOf<Element>} inputs
   * @param {Event} e
   */


  var handleResetEvent = (inputs, e) => {
    if (e.target.tagName === 'FORM' && e.target.classList.contains('form-material')) {
      delayDetectionAllInputs(inputs);
    }
  };
  /**
   * Setup forms fields listeners
   * @param {NodeListOf<Element>} inputElements
   */


  var setupFormsListeners = inputElements => {
    inputElements.forEach(input => input.firstInit = true);
    detectAllInputs(inputElements);
    var handleListenersRef = handleListeners.bind(null, inputElements);
    document.addEventListener('focus', handleListenersRef, true);
    document.addEventListener('blur', handleListenersRef, true);
    var delayDetectionAllInputsRef = delayDetectionAllInputs.bind(null, inputElements);
    window.addEventListener('pageshow', delayDetectionAllInputsRef);
    var handleResetRef = handleResetEvent.bind(null, inputElements);
    document.addEventListener('reset', handleResetRef);
  };
  /**
   * Update inputs state
   * @param {NodeListOf<Element>} inputElements
   */


  Axentix.updateInputs = function () {
    var inputElements = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : document.querySelectorAll('.form-material .form-field:not(.form-default) .form-control');
    var setupInputs = Array.from(inputElements).filter(el => !el.isInit);
    var detectInputs = Array.from(inputElements).filter(el => el.isInit);
    setupInputs.length > 0 ? setupFormsListeners(setupInputs) : '';
    detectInputs.length > 0 ? detectAllInputs(detectInputs) : '';
  };
})(); // Init


document.addEventListener('DOMContentLoaded', () => Axentix.updateInputs());

(() => {
  /**
   * Class lightbox
   * @class
   */
  class Lightbox extends AxentixComponent {
    static getDefaultOptions() {
      return {
        animationDuration: 400,
        overlayColor: 'grey dark-4',
        offset: 150,
        mobileOffset: 80,
        caption: ''
      };
    }
    /**
     * Construct Lightbox instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Lightbox',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Lightbox', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'lightbox.setup');
      this.el.style.transitionDuration = this.options.animationDuration + 'ms';
      this.container = Axentix.wrap([this.el]);

      this._setupListeners();
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.openOnClickRef = this.open.bind(this);
      this.el.addEventListener('click', this.openOnClickRef);
      this.closeEventRef = this.close.bind(this);
      window.addEventListener('keyup', this.closeEventRef);
      window.addEventListener('scroll', this.closeEventRef);
      window.addEventListener('resize', this.closeEventRef);
    }
    /**
     * Remove event listeners
     */


    _removeListeners() {
      this.el.removeEventListener('click', this.openOnClickRef);
      window.removeEventListener('keyup', this.closeEventRef);
      window.removeEventListener('scroll', this.closeEventRef);
      window.removeEventListener('resize', this.closeEventRef);
      this.openOnClickRef = undefined;
      this.onResizeRef = undefined;
      this.closeEventRef = undefined;
    }

    _setOverlay() {
      this.overlay = document.createElement('div');
      this.overlay.style.transitionDuration = this.options.animationDuration + 'ms';
      this.overlay.className = 'lightbox-overlay ' + this.options.overlayColor;
      this.container.appendChild(this.overlay);

      if (this.options.caption) {
        this.caption = document.createElement('p');
        this.caption.className = 'lightbox-caption';
        this.caption.innerHTML = this.options.caption;
        this.overlay.appendChild(this.caption);
      }

      this.overlayClickEventRef = this.close.bind(this);
      this.overlay.addEventListener('click', this.overlayClickEventRef);
    }

    _showOverlay() {
      this.overlay.style.opacity = 1;
    }

    _unsetOverlay() {
      this.overlay.style.opacity = 0;
      this.overlay.removeEventListener('click', this.overlayClickEventRef);
      setTimeout(() => {
        this.overlay.remove();
      }, this.options.animationDuration);
    }

    _calculateRatio() {
      var offset = window.innerWidth >= 960 ? this.options.offset : this.options.mobileOffset;

      if (window.innerWidth / window.innerHeight >= this.basicWidth / this.basicHeight) {
        this.newHeight = window.innerHeight - offset;
        this.newWidth = this.newHeight * this.basicWidth / this.basicHeight;
      } else {
        this.newWidth = window.innerWidth - offset;
        this.newHeight = this.newWidth * this.basicHeight / this.basicWidth;
      }
    }
    /**
     * Set position of active lightbox
     */


    open() {
      if (this.isActive) {
        this.close();
        return;
      } else if (this.isAnimated) {
        return;
      }

      var centerTop = window.innerHeight / 2;
      var centerLeft = window.innerWidth / 2;
      var rect = this.el.getBoundingClientRect();
      var containerRect = this.el.getBoundingClientRect();
      this.basicWidth = rect.width;
      this.el.style.width = this.basicWidth + 'px';
      this.basicHeight = rect.height;
      this.el.style.height = this.basicHeight + 'px';
      this.el.style.top = 0;
      this.el.style.left = 0;
      this.newTop = centerTop + window.scrollY - (containerRect.top + window.scrollY);
      this.newLeft = centerLeft + window.scrollX - (containerRect.left + window.scrollX);

      this._calculateRatio();

      this.container.style.position = 'relative';

      this._setOverlay();

      setTimeout(() => {
        Axentix.createEvent(this.el, 'lightbox.open');
        this.isAnimated = true;
        this.el.classList.add('active');

        if (this.el.classList.contains('responsive-media')) {
          this.el.classList.remove('responsive-media');
          this.isResponsive = true;
        } else {
          this.isResponsive = false;
        }

        this.isActive = true;

        this._showOverlay();

        this.container.style.width = this.basicWidth + 'px';
        this.container.style.height = this.basicHeight + 'px';
        this.el.style.width = this.newWidth + 'px';
        this.el.style.height = this.newHeight + 'px';
        this.el.style.top = this.newTop - this.newHeight / 2 + 'px';
        this.el.style.left = this.newLeft - this.newWidth / 2 + 'px';
        this.isAnimated = false;
      }, 50);
      setTimeout(() => {
        Axentix.createEvent(this.el, 'lightbox.opened');
      }, this.options.animationDuration + 50);
    }
    /**
     * Unset active lightbox
     */


    close(e) {
      if (!this.isActive || e && e.key && e.key !== 'Escape') {
        return;
      } else if (this.isAnimated) {
        return;
      }

      this.isAnimated = true;
      this.el.style.top = 0;
      this.el.style.left = 0;
      this.el.style.width = this.basicWidth + 'px';
      this.el.style.height = this.basicHeight + 'px';

      this._unsetOverlay();

      Axentix.createEvent(this.el, 'lightbox.close');
      setTimeout(() => {
        this.el.classList.remove('active');
        this.isResponsive ? this.el.classList.add('responsive-media') : '';
        this.container.removeAttribute('style');
        this.el.style.left = '';
        this.el.style.top = '';
        this.el.style.width = '';
        this.el.style.height = '';
        this.el.style.transform = '';
        this.isActive = false;
        this.isAnimated = false;
        Axentix.createEvent(this.el, 'lightbox.closed');
      }, this.options.animationDuration + 50);
    }

  }

  Axentix.Config.registerComponent({
    class: Lightbox,
    name: 'Lightbox',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.lightbox:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Modal
   * @class
   */
  class Modal extends AxentixComponent {
    static getDefaultOptions() {
      return {
        overlay: true,
        bodyScrolling: false,
        animationDuration: 400
      };
    }
    /**
     * Construct Modal instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Modal',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Modal', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'modal.setup');
      this.modalTriggers = document.querySelectorAll('.modal-trigger');
      this.isActive = this.el.classList.contains('active') ? true : false;
      this.isAnimated = false;

      this._setupListeners();

      this.options.overlay ? this._createOverlay() : '';
      this.el.style.transitionDuration = this.options.animationDuration + 'ms';
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.listenerRef = this._onClickTrigger.bind(this);
      this.modalTriggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          trigger.addEventListener('click', this.listenerRef);
        }
      });
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      this.modalTriggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          trigger.removeEventListener('click', this.listenerRef);
        }
      });
      this.listenerRef = undefined;
    }
    /**
     * Create overlay element
     */


    _createOverlay() {
      if (this.isActive && this.options.overlay) {
        this.overlayElement = document.querySelector('.modal-overlay[data-target="' + this.el.id + '"]');
        this.overlayElement ? '' : this.overlayElement = document.createElement('div');
      } else {
        this.overlayElement = document.createElement('div');
      }

      this.overlayElement.classList.add('modal-overlay');
      this.overlayElement.style.transitionDuration = this.options.animationDuration + 'ms';
      this.overlayElement.dataset.target = this.el.id;
    }
    /**
     * Enable or disable body scroll when option is true
     * @param {boolean} state Enable or disable body scroll
     */


    _toggleBodyScroll(state) {
      if (!this.options.bodyScrolling) {
        state ? document.body.style.overflow = '' : document.body.style.overflow = 'hidden';
      }
    }
    /**
     * Set Z-Index when modal is open
     */


    _setZIndex() {
      var totalModals = document.querySelectorAll('.modal.active').length + 1;
      this.options.overlay ? this.overlayElement.style.zIndex = 800 + totalModals * 6 : '';
      this.el.style.zIndex = 800 + totalModals * 10;
    }
    /**
     * Handle click on trigger
     */


    _onClickTrigger(e) {
      e.preventDefault();

      if (this.isAnimated) {
        return;
      }

      this.isActive ? this.close() : this.open();
    }
    /**
     * Open the modal
     */


    open() {
      if (this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'modal.open');
      this.isActive = true;
      this.isAnimated = true;

      this._setZIndex();

      this.el.style.display = 'block';
      this.overlay(true);

      this._toggleBodyScroll(false);

      setTimeout(() => {
        this.el.classList.add('active');
      }, 50);
      setTimeout(() => {
        this.isAnimated = false;
        Axentix.createEvent(this.el, 'modal.opened');
      }, this.options.animationDuration);
    }
    /**
     * Close the modal
     */


    close() {
      if (!this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'modal.close');
      this.isAnimated = true;
      this.el.classList.remove('active');
      this.overlay(false);
      setTimeout(() => {
        this.el.style.display = '';
        this.isAnimated = false;
        this.isActive = false;

        this._toggleBodyScroll(true);

        Axentix.createEvent(this.el, 'modal.closed');
      }, this.options.animationDuration);
    }
    /**
     * Manage overlay
     * @param {boolean} state
     */


    overlay(state) {
      if (this.options.overlay) {
        if (state) {
          this.overlayElement.addEventListener('click', this.listenerRef);
          document.body.appendChild(this.overlayElement);
          setTimeout(() => {
            this.overlayElement.classList.add('active');
          }, 50);
        } else {
          this.overlayElement.classList.remove('active');
          setTimeout(() => {
            this.overlayElement.removeEventListener('click', this.listenerRef);
            document.body.removeChild(this.overlayElement);
          }, this.options.animationDuration);
        }
      }
    }

  }

  Axentix.Config.registerComponent({
    class: Modal,
    name: 'Modal',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.modal:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class ScrollSpy
   * @class
   */
  class ScrollSpy extends AxentixComponent {
    static getDefaultOptions() {
      return {
        offset: 200,
        linkSelector: 'a',
        classes: 'active',
        auto: {
          enabled: false,
          classes: '',
          selector: ''
        }
      };
    }
    /**
     * Construct ScrollSpy instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'ScrollSpy',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('ScrollSpy', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'scrollspy.setup');
      this.options.auto.enabled ? this._setupAuto() : this._setupBasic();
      this.options.classes = this.options.classes.split(' ');
      this.oldLink = '';

      this._setupListeners();

      this._update();
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.updateRef = this._update.bind(this);
      window.addEventListener('scroll', this.updateRef);
      window.addEventListener('resize', this.updateRef);
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      window.removeEventListener('scroll', this.updateRef);
      window.removeEventListener('resize', this.updateRef);
      this.updateRef = undefined;
    }

    _setupBasic() {
      this.links = Array.from(this.el.querySelectorAll(this.options.linkSelector));
      this.elements = this.links.map(link => document.querySelector(link.getAttribute('href')));
    }

    _setupAuto() {
      this.elements = Array.from(document.querySelectorAll(this.options.auto.selector));
      this.links = this.elements.map(el => {
        var link = document.createElement('a');
        link.className = this.options.auto.classes;
        link.setAttribute('href', '#' + el.id);
        link.innerHTML = el.innerHTML;
        this.el.appendChild(link);
        return link;
      });
    }

    _getElement() {
      var top = window.scrollY,
          left = window.scrollX,
          right = window.innerWidth,
          bottom = window.innerHeight,
          topBreakpoint = top + this.options.offset;

      if (bottom + top >= document.body.offsetHeight - 2) {
        return this.elements[this.elements.length - 1];
      }

      return this.elements.find(el => {
        var elRect = el.getBoundingClientRect();
        return elRect.top + top >= top && elRect.left + left >= left && elRect.right <= right && elRect.bottom <= bottom && elRect.top + top <= topBreakpoint;
      });
    }

    _removeOldLink() {
      if (!this.oldLink) {
        return;
      }

      this.options.classes.map(cl => this.oldLink.classList.remove(cl));
    }

    _getClosestElem() {
      var top = window.scrollY;
      return this.elements.reduce((prev, curr) => {
        var currTop = curr.getBoundingClientRect().top + top;
        var prevTop = prev.getBoundingClientRect().top + top;
        return currTop > top + this.options.offset ? prev : Math.abs(currTop - top) < Math.abs(prevTop - top) ? curr : prev;
      });
    }

    _update() {
      var element = this._getElement();

      element ? '' : element = this._getClosestElem();
      var link = this.links.find(link => link.getAttribute('href').split('#')[1] === element.id);

      if (link === this.oldLink) {
        return;
      }

      Axentix.createEvent(this.el, 'scrollspy.update');

      this._removeOldLink();

      this.options.classes.map(cl => link.classList.add(cl));
      this.oldLink = link;
    }

  }

  Axentix.Config.registerComponent({
    class: ScrollSpy,
    name: 'ScrollSpy',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.scrollspy:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Sidenav
   * @class
   */
  class Sidenav extends AxentixComponent {
    static getDefaultOptions() {
      return {
        overlay: true,
        bodyScrolling: false,
        animationDuration: 300
      };
    }
    /**
     * Construct Sidenav instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Sidenav',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Sidenav', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'sidenav.setup');
      this.sidenavTriggers = document.querySelectorAll('.sidenav-trigger');
      this.isActive = false;
      this.isFixed = this.el.classList.contains('fixed');
      this.isLarge = this.el.classList.contains('large');
      this.layoutEl = document.querySelector('.layout');

      this._setupListeners();

      this.options.overlay ? this._createOverlay() : '';
      this.el.classList.contains('large') && this.layoutEl ? this.layoutEl.classList.add('sidenav-large') : this.layoutEl.classList.remove('sidenav-large');

      this._handleRightSidenav();

      this.el.style.transitionDuration = this.options.animationDuration + 'ms';
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.listenerRef = this._onClickTrigger.bind(this);
      this.sidenavTriggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          trigger.addEventListener('click', this.listenerRef);
        }
      });
      this.windowResizeRef = this.close.bind(this);
      window.addEventListener('resize', this.windowResizeRef);
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      this.sidenavTriggers.forEach(trigger => {
        if (trigger.dataset.target === this.el.id) {
          trigger.removeEventListener('click', this.listenerRef);
        }
      });
      this.listenerRef = undefined;
      window.removeEventListener('resize', this.windowResizeRef);
      this.windowResizeRef = undefined;
    }
    /**
     * Handle right sidenav detection
     */


    _handleRightSidenav() {
      var sidenavs = document.querySelectorAll('.sidenav');
      var found = Array.from(sidenavs).some(sidenav => sidenav.classList.contains('right-aligned'));

      if (found && this.layoutEl && !this.layoutEl.classList.contains('sidenav-right')) {
        this.layoutEl.classList.add('sidenav-right');
      } else if (!found && this.layoutEl && this.layoutEl.classList.contains('sidenav-right')) {
        this.layoutEl.classList.remove('sidenav-right');
      }
    }
    /**
     * Create overlay element
     */


    _createOverlay() {
      this.overlayElement = document.createElement('div');
      this.overlayElement.classList.add('sidenav-overlay');
      this.overlayElement.dataset.target = this.el.id;
    }
    /**
     * Enable or disable body scroll when option is true
     * @param {boolean} state
     */


    _toggleBodyScroll(state) {
      if (!this.options.bodyScrolling) {
        state ? document.body.style.overflow = '' : document.body.style.overflow = 'hidden';
      }
    }
    /**
     * Handle click on trigger
     * @param {Event} e
     */


    _onClickTrigger(e) {
      e.preventDefault();

      if (this.isFixed && window.innerWidth >= 960) {
        return;
      }

      this.isActive ? this.close() : this.open();
    }
    /**
     * Open sidenav
     */


    open() {
      if (this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'sidenav.open');
      this.isActive = true;
      this.el.classList.add('active');
      this.overlay(true);

      this._toggleBodyScroll(false);

      setTimeout(() => {
        Axentix.createEvent(this.el, 'sidenav.opened');
      }, this.options.animationDuration);
    }
    /**
     * Close sidenav
     */


    close() {
      if (!this.isActive) {
        return;
      }

      Axentix.createEvent(this.el, 'sidenav.close');
      this.el.classList.remove('active');
      this.overlay(false);
      setTimeout(() => {
        this._toggleBodyScroll(true);

        this.isActive = false;
        Axentix.createEvent(this.el, 'sidenav.closed');
      }, this.options.animationDuration);
    }
    /**
     * Manage overlay
     * @param {boolean} state
     */


    overlay(state) {
      if (this.options.overlay) {
        if (state) {
          this.overlayElement.addEventListener('click', this.listenerRef);
          document.body.appendChild(this.overlayElement);
        } else {
          this.overlayElement.removeEventListener('click', this.listenerRef);
          document.body.removeChild(this.overlayElement);
        }
      }
    }

  }

  Axentix.Config.registerComponent({
    class: Sidenav,
    name: 'Sidenav',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.sidenav:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Tab
   * @class
   */
  class Tab extends AxentixComponent {
    static getDefaultOptions() {
      return {
        animationDuration: 300,
        animationType: 'none',
        disableActiveBar: false,
        caroulix: {}
      };
    }
    /**
     * Construct Tab instance
     * @constructor
     * @param {String} element
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Tab',
        instance: this
      });
      this.caroulixOptions = {
        animationDuration: 300,
        autoplay: {
          enabled: false
        }
      };
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Tab', options, this.el, isLoadedWithData);

      this._setup();
    }
    /**
     * Setup component
     */


    _setup() {
      Axentix.createEvent(this.el, 'tab.setup');
      var animationList = ['none', 'slide'];
      animationList.includes(this.options.animationType) ? '' : this.options.animationType = 'none';
      this.isAnimated = false;
      this.tabArrow = this.el.querySelector('.tab-arrow');
      this.tabLinks = this.el.querySelectorAll('.tab-menu .tab-link');
      this.tabMenu = this.el.querySelector('.tab-menu');
      this.currentItemIndex = 0;

      this._getItems();

      if (this.tabArrow) {
        this._toggleArrowMode();

        this.leftArrow = this.el.querySelector('.tab-arrow .tab-prev');
        this.rightArrow = this.el.querySelector('.tab-arrow .tab-next');
      }

      this._setupListeners();

      this.el.style.transitionDuration = this.options.animationDuration + 'ms';
      this.options.animationType === 'slide' ? this._enableSlideAnimation() : this.updateActiveElement();
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.tabLinks.forEach(item => {
        item.listenerRef = this._onClickItem.bind(this, item);
        item.addEventListener('click', item.listenerRef);
      });
      this.resizeTabListener = this._handleResizeEvent.bind(this);
      window.addEventListener('resize', this.resizeTabListener);

      if (this.tabArrow) {
        this.arrowListener = this._toggleArrowMode.bind(this);
        window.addEventListener('resize', this.arrowListener);
        this.scrollLeftListener = this._scrollLeft.bind(this);
        this.scrollRightLstener = this._scrollRight.bind(this);
        this.leftArrow.addEventListener('click', this.scrollLeftListener);
        this.rightArrow.addEventListener('click', this.scrollRightLstener);
      }
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      this.tabLinks.forEach(item => {
        item.removeEventListener('click', item.listenerRef);
        item.listenerRef = undefined;
      });
      window.removeEventListener('resize', this.resizeTabListener);
      this.resizeTabListener = undefined;

      if (this.tabArrow) {
        window.removeEventListener('resize', this.arrowListener);
        this.arrowListener = undefined;
        this.leftArrow.removeEventListener('click', this.scrollLeftListener);
        this.rightArrow.removeEventListener('click', this.scrollRightLstener);
        this.scrollLeftListener = undefined;
        this.scrollRightLstener = undefined;
      }
    }

    _handleResizeEvent() {
      this.updateActiveElement();

      for (var i = 100; i < 500; i += 100) {
        setTimeout(() => {
          this.updateActiveElement();
        }, i);
      }
    }
    /**
     * Get all items
     */


    _getItems() {
      this.tabItems = Array.from(this.el.children).reduce((acc, child) => {
        !child.classList.contains('tab-menu') && !child.classList.contains('tab-arrow') ? acc.push(child) : '';
        return acc;
      }, []);
    }
    /**
     * Hide all tab items
     */


    _hideContent() {
      this.tabItems.map(item => item.style.display = 'none');
    }
    /**
     * Init slide animation
     */


    _enableSlideAnimation() {
      this.tabItems.map(item => item.classList.add('caroulix-item'));
      this.tabCaroulix = Axentix.wrap(this.tabItems);
      this.tabCaroulix.classList.add('caroulix');
      var nb = Math.random().toString().split('.')[1];
      this.tabCaroulix.id = 'tab-caroulix-' + nb;
      this.tabCaroulixInit = true;
      this.options.caroulix = Axentix.extend(this.caroulixOptions, this.options.caroulix);
      this.options.animationDuration !== 300 ? this.options.caroulix.animationDuration = this.options.animationDuration : '';
      this.updateActiveElement();
    }
    /**
     * Set active bar position
     * @param {Element} element
     */


    _setActiveElement(element) {
      this.tabLinks.forEach(item => item.classList.remove('active'));

      if (!this.options.disableActiveBar) {
        var elementRect = element.getBoundingClientRect();
        var elementPosLeft = elementRect.left;
        var menuPosLeft = this.tabMenu.getBoundingClientRect().left;
        var left = elementPosLeft - menuPosLeft + this.tabMenu.scrollLeft;
        var elementWidth = elementRect.width;
        var right = this.tabMenu.clientWidth - left - elementWidth;
        this.tabMenu.style.setProperty('--tab-bar-left-offset', Math.floor(left) + 'px');
        this.tabMenu.style.setProperty('--tab-bar-right-offset', Math.ceil(right) + 'px');
      }

      element.classList.add('active');
    }
    /**
     * Toggle arrow mode
     */


    _toggleArrowMode() {
      var totalWidth = Array.from(this.tabLinks).reduce((acc, element) => {
        acc += element.clientWidth;
        return acc;
      }, 0);
      var arrowWidth = this.tabArrow.clientWidth;

      if (totalWidth > arrowWidth) {
        this.tabArrow.classList.contains('tab-arrow-show') ? '' : this.tabArrow.classList.add('tab-arrow-show');
      } else {
        this.tabArrow.classList.contains('tab-arrow-show') ? this.tabArrow.classList.remove('tab-arrow-show') : '';
      }

      this.updateActiveElement();
    }
    /**
     * Scroll left
     * @param {Event} e
     */


    _scrollLeft(e) {
      e.preventDefault();
      this.tabMenu.scrollLeft -= 40;
    }
    /**
     * Scroll right
     * @param {Event} e
     */


    _scrollRight(e) {
      e.preventDefault();
      this.tabMenu.scrollLeft += 40;
    }
    /**
     * Handle click on menu item
     * @param {Element} item
     * @param {Event} e
     */


    _onClickItem(item, e) {
      e.preventDefault();

      if (this.isAnimated || item.classList.contains('active')) {
        return;
      }

      var target = item.children[0].getAttribute('href');
      this.select(target.split('#')[1]);
    }

    _getPreviousItemIndex(step) {
      var previousItemIndex = 0;
      var index = this.currentItemIndex;

      for (var i = 0; i < step; i++) {
        if (index > 0) {
          previousItemIndex = index - 1;
          index--;
        } else {
          index = this.tabLinks.length - 1;
          previousItemIndex = index;
        }
      }

      return previousItemIndex;
    }

    _getNextItemIndex(step) {
      var nextItemIndex = 0;
      var index = this.currentItemIndex;

      for (var i = 0; i < step; i++) {
        if (index < this.tabLinks.length - 1) {
          nextItemIndex = index + 1;
          index++;
        } else {
          index = 0;
          nextItemIndex = index;
        }
      }

      return nextItemIndex;
    }
    /**
     * Select tab
     * @param {String} itemId
     */


    select(itemId) {
      if (this.isAnimated) {
        return;
      }

      this.isAnimated = true;
      var menuItem = this.el.querySelector('.tab-menu a[href="#' + itemId + '"]');
      this.currentItemIndex = Array.from(this.tabLinks).findIndex(item => item.children[0] === menuItem);
      Axentix.createEvent(menuItem, 'tab.select', {
        currentIndex: this.currentItemIndex
      });

      this._setActiveElement(menuItem.parentElement);

      if (this.tabCaroulixInit) {
        this.tabItems.map(item => item.id === itemId ? item.classList.add('active') : '');
        this.caroulixInstance = new Axentix.Caroulix('#' + this.tabCaroulix.id, this.options.caroulix, this.el, true);
        this.tabCaroulixInit = false;
        this.isAnimated = false;
        return;
      }

      if (this.options.animationType === 'slide') {
        var nb = this.tabItems.findIndex(item => item.id === itemId);
        this.caroulixInstance.goTo(nb);
        setTimeout(() => {
          this.isAnimated = false;
        }, this.options.animationDuration);
      } else {
        this._hideContent();

        this.tabItems.map(item => item.id === itemId ? item.style.display = 'block' : '');
        this.isAnimated = false;
      }
    }
    /**
     * Detect active element & update component
     */


    updateActiveElement() {
      var itemSelected;
      this.tabLinks.forEach((item, index) => {
        item.classList.contains('active') ? (itemSelected = item, this.currentItemIndex = index) : '';
      });
      itemSelected ? '' : (itemSelected = this.tabLinks.item(0), this.currentItemIndex = 0);
      var target = itemSelected.children[0].getAttribute('href');
      this.tabSelected = target;
      this.select(target.split('#')[1]);
    }
    /**
     * Go to previous tab
     */


    prev() {
      var step = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

      if (this.isAnimated) {
        return;
      }

      var previousItemIndex = this._getPreviousItemIndex(step);

      this.currentItemIndex = previousItemIndex;
      Axentix.createEvent(this.el, 'tab.prev', {
        step
      });
      var target = this.tabLinks[previousItemIndex].children[0].getAttribute('href');
      this.select(target.split('#')[1]);
    }
    /**
     * Go to next tab
     */


    next() {
      var step = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

      if (this.isAnimated) {
        return;
      }

      var nextItemIndex = this._getNextItemIndex(step);

      this.currentItemIndex = nextItemIndex;
      Axentix.createEvent(this.el, 'tab.next', {
        step
      });
      var target = this.tabLinks[nextItemIndex].children[0].getAttribute('href');
      this.select(target.split('#')[1]);
    }

  }

  Axentix.Config.registerComponent({
    class: Tab,
    name: 'Tab',
    dataDetection: true,
    autoInit: {
      enabled: true,
      selector: '.tab:not(.no-axentix-init)'
    }
  });
})();

(() => {
  /**
   * Class Toast
   * @class
   */
  class Toast {
    static getDefaultOptions() {
      return {
        animationDuration: 400,
        duration: 4000,
        classes: '',
        position: 'right',
        direction: 'top',
        mobileDirection: 'bottom',
        isClosable: false
      };
    }
    /**
     * Construct Toast instance
     * @constructor
     * @param {String} content
     * @param {Object} options
     */


    constructor(content, options) {
      if (Axentix.toastInstanceExist) {
        console.error("Don't try to create multiple toast instances");
        return;
      } else {
        Axentix.toastInstanceExist = true;
      }

      Axentix.instances.push({
        type: 'Toast',
        instance: this
      });
      this.content = content;
      this.options = Axentix.getComponentOptions('Toast', options, '', true);
      this.options.position = this.options.position.toLowerCase();
      this.options.direction = this.options.direction.toLowerCase();
      this.options.mobileDirection = this.options.mobileDirection.toLowerCase();
      this.toasters = {};
    }
    /**
     * Create toast container
     */


    _createToaster() {
      var toaster = document.createElement('div');
      var positionList = ['right', 'left'];
      positionList.includes(this.options.position) ? '' : this.options.position = 'right';
      var directionList = ['bottom', 'top'];
      directionList.includes(this.options.direction) ? '' : this.options.direction = 'top';
      directionList.includes(this.options.mobileDirection) ? '' : this.options.mobileDirection = 'bottom';
      toaster.className = 'toaster toaster-' + this.options.position + ' toast-' + this.options.direction + ' toaster-mobile-' + this.options.mobileDirection;
      this.toasters[this.options.position] = toaster;
      document.body.appendChild(toaster);
    }
    /**
     * Remove toast container
     */


    _removeToaster() {
      for (var key in this.toasters) {
        var toaster = this.toasters[key];

        if (toaster.childElementCount <= 0) {
          toaster.remove();
          delete this.toasters[key];
        }
      }
    }
    /**
     * Toast in animation
     * @param {Element} toast
     */


    _fadeInToast(toast) {
      setTimeout(() => {
        Axentix.createEvent(toast, 'toast.show');
        toast.classList.add('toast-animated');
        setTimeout(() => {
          Axentix.createEvent(toast, 'toast.shown');
        }, this.options.animationDuration);
      }, 50);
    }
    /**
     * Toast out animation
     * @param {Element} toast
     */


    _fadeOutToast(toast) {
      setTimeout(() => {
        Axentix.createEvent(toast, 'toast.hide');

        this._hide(toast);
      }, this.options.duration + this.options.animationDuration);
    }
    /**
     * Anim out toast
     * @param {Element} toast
     */


    _animOut(toast) {
      toast.style.transitionTimingFunction = 'cubic-bezier(0.445, 0.05, 0.55, 0.95)';
      toast.style.paddingTop = 0;
      toast.style.paddingBottom = 0;
      toast.style.margin = 0;
      toast.style.height = 0;
    }
    /**
     * Create toast
     */


    _createToast() {
      var toast = document.createElement('div');
      toast.className = 'toast shadow-1 ' + this.options.classes;
      toast.innerHTML = this.content;
      toast.style.transitionDuration = this.options.animationDuration + 'ms';

      if (this.options.isClosable) {
        var trigger = document.createElement('i');
        trigger.className = 'toast-trigger fas fa-times';
        trigger.listenerRef = this._hide.bind(this, toast, trigger);
        trigger.addEventListener('click', trigger.listenerRef);
        toast.appendChild(trigger);
      }

      this._fadeInToast(toast);

      this.toasters[this.options.position].appendChild(toast);

      this._fadeOutToast(toast);

      var height = toast.clientHeight;
      toast.style.height = height + 'px';
    }
    /**
     * Hide toast
     * @param {String} toast
     * @param {Element} trigger
     * @param {Event} e
     */


    _hide(toast, trigger, e) {
      if (toast.isAnimated) {
        return;
      }

      var timer = 1;

      if (e) {
        e.preventDefault();
        timer = 0;
        this.options.isClosable ? trigger.removeEventListener('click', trigger.listenerRef) : '';
      }

      toast.style.opacity = 0;
      toast.isAnimated = true;
      var delay = timer * this.options.animationDuration + this.options.animationDuration;
      setTimeout(() => {
        this._animOut(toast);
      }, delay / 2);
      setTimeout(() => {
        toast.remove();
        Axentix.createEvent(toast, 'toast.remove');

        this._removeToaster();
      }, delay * 1.45);
    }
    /**
     * Showing the toast
     */


    show() {
      if (!Object.keys(this.toasters).includes(this.options.position)) {
        this._createToaster();
      }

      this._createToast();
    }
    /**
     * Change
     * @param {String} content
     * @param {Object} options
     */


    change(content, options) {
      this.content = content;
      this.options = Axentix.extend(this.options, options);
    }

  }

  Axentix.Config.registerComponent({
    class: Toast,
    name: 'Toast'
  });
})();

(() => {
  /**
   * Class Tooltip
   * @Class
   */
  class Tooltip extends AxentixComponent {
    static getDefaultOptions() {
      return {
        content: '',
        animationDelay: 0,
        offset: '10px',
        animationDuration: 200,
        classes: 'grey dark-4 light-shadow-2 p-2',
        position: 'top'
      };
    }
    /**
     * Tooltip constructor
     * @constructor
     * @param {Object} options
     */


    constructor(element, options, isLoadedWithData) {
      super();
      Axentix.instances.push({
        type: 'Tooltip',
        instance: this
      });
      this.el = document.querySelector(element);
      this.options = Axentix.getComponentOptions('Tooltip', options, this.el, isLoadedWithData);

      this._setup();
    }

    _setup() {
      if (!this.options.content) {
        return console.error('Tooltip #' + this.el.id + ' : empty content.');
      }

      Axentix.createEvent(this.el, 'tooltip.setup');
      this.options.position = this.options.position.toLowerCase();
      var tooltips = document.querySelectorAll('.tooltip');
      tooltips.forEach(tooltip => {
        tooltip.dataset.tooltipId ? tooltip.dataset.tooltipId === this.el.id ? this.tooltip = tooltip : '' : '';
      });
      this.tooltip ? '' : this.tooltip = document.createElement('div');
      this.tooltip.dataset.tooltipId === this.el.id ? '' : this.tooltip.dataset.tooltipId = this.el.id;

      this._setProperties();

      document.body.appendChild(this.tooltip);
      this.positionList = ['right', 'left', 'top', 'bottom'];
      this.positionList.includes(this.options.position) ? '' : this.options.position = 'top';

      this._setupListeners();

      this.updatePosition();
    }
    /**
     * Setup listeners
     */


    _setupListeners() {
      this.listenerEnterRef = this._onHover.bind(this);
      this.listenerLeaveRef = this._onHoverOut.bind(this);
      this.el.addEventListener('mouseenter', this.listenerEnterRef);
      this.el.addEventListener('mouseleave', this.listenerLeaveRef);
    }
    /**
     * Remove listeners
     */


    _removeListeners() {
      this.el.removeEventListener('mouseenter', this.listenerEnterRef);
      this.el.removeEventListener('mouseleave', this.listenerLeaveRef);
      this.listenerEnterRef = undefined;
      this.listenerLeaveRef = undefined;
    }
    /**
     * Set properties to tooltip
     */


    _setProperties() {
      this.tooltip.style.transform = 'translate(0)';
      this.tooltip.style.opacity = 0;
      this.tooltip.className = 'tooltip ' + this.options.classes;
      this.tooltip.style.transitionDuration = this.options.animationDuration + 'ms';
      this.tooltip.innerHTML = this.options.content;
    }
    /**
     * Set basic tooltip position
     */


    _setBasicPosition() {
      if (this.options.position == 'top' || this.options.position == 'bottom') {
        this.options.position == 'top' ? this.tooltip.style.top = this.elRect.top + 'px' : this.tooltip.style.top = this.elRect.top + this.elRect.height + 'px';
      } else if (this.options.position == 'left' || this.options.position == 'right') {
        this.options.position == 'right' ? this.tooltip.style.left = this.elRect.left + this.elRect.width + 'px' : '';
      }
    }
    /**
     * Manually transform the tooltip location
     */


    _manualTransform() {
      if (this.options.position == 'top' || this.options.position == 'bottom') {
        this.tooltip.style.left = this.elRect.left + this.elRect.width / 2 - this.tooltipRect.width / 2 + 'px';
      } else if (this.options.position == 'left' || this.options.position == 'right') {
        this.tooltip.style.top = this.elRect.top + this.elRect.height / 2 - this.tooltipRect.height / 2 + 'px';
      }

      if (this.options.position == 'top') {
        this.tooltip.style.top = this.tooltipRect.top - this.tooltipRect.height + 'px';
      } else if (this.options.position == 'left') {
        this.tooltip.style.left = this.elRect.left - this.tooltipRect.width + 'px';
      }

      var scrollY = window.scrollY;
      var tooltipTop = parseFloat(this.tooltip.style.top);
      this.options.position === 'top' ? this.tooltip.style.top = scrollY * 2 + tooltipTop + 'px' : this.tooltip.style.top = scrollY + tooltipTop + 'px';
    }
    /**
     * Handle hover event
     * @param {Event} e
     */


    _onHover(e) {
      e.preventDefault();
      this.show();
    }
    /**
     * Handle hover event
     * @param {Event} e
     */


    _onHoverOut(e) {
      e.preventDefault();
      this.hide();
    }
    /**
     * Update current tooltip position
     */


    updatePosition() {
      this.elRect = this.el.getBoundingClientRect();

      this._setBasicPosition();

      this.tooltipRect = this.tooltip.getBoundingClientRect();

      this._manualTransform();
    }
    /**
     * Show tooltip
     */


    show() {
      this.updatePosition();
      setTimeout(() => {
        Axentix.createEvent(this.el, 'tooltip.show');
        this.options.position == 'top' ? this.tooltip.style.transform = "translateY(-".concat(this.options.offset, ")") : this.options.position == 'right' ? this.tooltip.style.transform = "translateX(".concat(this.options.offset, ")") : this.options.position == 'bottom' ? this.tooltip.style.transform = "translateY(".concat(this.options.offset, ")") : this.options.position == 'left' ? this.tooltip.style.transform = "translateX(-".concat(this.options.offset, ")") : '';
        this.tooltip.style.opacity = 1;
      }, this.options.animationDelay);
    }
    /**
     * Hide tooltip
     */


    hide() {
      Axentix.createEvent(this.el, 'tooltip.hide');
      this.tooltip.style.transform = 'translate(0)';
      this.tooltip.style.opacity = 0;
    }
    /**
     * Change current options
     * @param {Object} options
     */


    change() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      this.options = Axentix.getComponentOptions('Tooltip', options, this.el, true);
      this.positionList.includes(this.options.position) ? '' : this.options.position = 'top';

      this._setProperties();

      this.updatePosition();
    }

  }

  Axentix.Config.registerComponent({
    class: Tooltip,
    name: 'Tooltip',
    dataDetection: true
  });
})();

Axentix.extend = function () {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args.reduce((acc, obj) => {
    for (var key in obj) {
      typeof obj[key] === 'object' && obj[key] !== null ? acc[key] = Axentix.extend(acc[key], obj[key]) : acc[key] = obj[key];
    }

    return acc;
  }, {});
};

Axentix.getComponentOptions = (component, options, el, isLoadedWithData) => {
  return Axentix.extend(Axentix[component].getDefaultOptions(), isLoadedWithData ? {} : Axentix.DataDetection.formatOptions(component, el), options);
};
/**
 * Wrap content inside an element (<div> by default)
 * @param {Array<Element>} target
 * @param {Element} wrapper
 * @return {Element}
 */


Axentix.wrap = function (target) {
  var wrapper = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : document.createElement('div');
  var parent = target[0].parentElement;
  parent.insertBefore(wrapper, target[0]);
  target.forEach(elem => wrapper.appendChild(elem));
  return wrapper;
};
/**
 * Create custom event
 * @param {Element} element
 * @param {string} eventName
 * @param {Object} extraData
 */


Axentix.createEvent = (element, eventName, extraData) => {
  var event = new CustomEvent('ax.' + eventName, {
    detail: extraData || {},
    bubbles: true
  });
  element.dispatchEvent(event);
};

Axentix.isTouchEnabled = () => {
  return 'ontouchstart' in window || navigator.maxTouchPoints > 0 || navigator.msMaxTouchPoints > 0;
};

Axentix.getInstanceByType = type => {
  return Axentix.instances.filter(ins => ins.type === type).map(ins => ins.instance);
};

Axentix.getInstance = element => {
  var el = Axentix.instances.find(ins => '#' + ins.instance.el.id === element);

  if (el) {
    return el.instance;
  }

  return false;
};

Axentix.getAllInstances = () => {
  return Axentix.instances;
};

Axentix.sync = element => {
  Axentix.getInstance(element).sync();
};

Axentix.syncAll = () => {
  Axentix.instances.map(ins => ins.instance.sync());
};

Axentix.reset = element => {
  Axentix.getInstance(element).reset();
};

Axentix.resetAll = () => {
  Axentix.instances.map(ins => ins.instance.reset());
};
